enhancements/installer/gcp-ipi-shared-vpc.md

```diff
@@ -0,0 +1,282 @@
+  minimal permission requirements in the Host Project.
+
+### Non-Goals
+
+- Provisioning and/or configuring the VPC, Network, and Subnetworks are to be
+  done manually by the user prior to running the installer.
+- Provisioning and/or configuring the Service Accounts for the Installer and
+  Cluster are to be done manually by the user prior to running the installer.
+- Providing solution(s) for manually managing DNS in the Host Project is to be
+  solved by another effort.
+
+## Proposal
+
+The installer should be able to install a GCP IPI cluster using Shared VPC
+networking by automating the majority of the current GCP UPI workflow. This
+requires some resources to be created and configured in advance, such as the
+network and subnets. Other resources can be created in advance when the
+necessary host project permissions are not desired, such as the firewall rules
+and/or DNS configuration. The installer, and the cluster-ingress-operator, will
+be updated to enable some of this functionality. They will become capable of
+creating and managing resources that exist in a different (Host) project from
+the cluster (Service Project). They will also become capable of consuming
+pre-existing firewall rules.
+
+### Workflow Description
+
+Prior to executing the installer, the Shared VPC, Network, and Subnets will be
+manually created in the Host Project and configured for Shared VPC (XPN)
+configuration with the Service Project. A service account will also be created
+in the Service Project and granted sufficient permissions in the Host Project.
+When minimal permissions are desired in the Host Project, additional resources
+will be manually created and configured to include: firewall rules, api compute
+address, ingress compute address, and the api, api-int, and \*.apps public
+and/or private DNS records.
+
+After all of these resources are created, they will be used in a manually
+created install-config.yaml to configure the cluster accordingly. At a minimum,
+the credentialsMode will be set to Passthrough, and the network,
+computeNetwork, controlPlaneNetwork, and networkProjectID will be configured to
+reference the Shared VPC network. When minimial permissions are desired in the
+Host Project, the additional parameters will be configured to enable the
+cluster to consume them as well.
+
+The installation process can then proceed as any other IPI workflow by running
+the create manifests, create ignition, and/or create cluster sub-commands.
+
+### API Extensions
+
+#### api.dns
+
+The current DNSZone struct does not appear to be able to store which project
+the zone is in. The ID variable is currently limited to a managedZone ID value.
+The cluster-ingress-operator will be updated to also accept a relative resource
+name in the following format. This will enable the operator to override the
+default project id with the one provided. No functional API changes are
+necessary.
+
+projects/{projectid}/managedZones/{zoneid}
+
```

Comment:
The installer will update the `cluster-dns-02-config.yml` (ie DNS:cluster) to use this format for the spec.privateZone and spec.publicZone parameters, only when a projectid is provided that is different from the clusterid. Otherwise it will remain using the current format of just the zoneID.

And yes, the format is to enable specifying a projectID for the zone specified by zoneID. This is inline with the gcp definition for a resource name, specifically a `Relative Resource Name`
https://cloud.google.com/apis/design/resource_names

Reply:
An example `cluster-dns-02-config.yaml` is as follows. In the example, you will see that the private zone is using the original syntax of a simple zoneID. This zone is being created by the installer as the `privateDNSZone` parameter was omitted in the `install-config.yaml`. On the other hand, the public zone is using the newer relative resource name syntax. The `publicDNSZone` was specified in the install-config.yaml. with `{ ID: my-public-zone, projectID: myproject }`

```
apiVersion: config.openshift.io/v1
kind: DNS
metadata:
  creationTimestamp: null
  name: cluster
spec:
  baseDomain: cluster.example.com
  privateZone:
    id: cluster-zpqfj-private-zone
  publicZone:
    id: project/myproject/managedZones/my-public-zone
status: {}
```

Reply:
An example of how this could be handled by the cluster-ingress-operator can be found in the following PR
https://github.com/openshift/cluster-ingress-operator/pull/827

Reply:
The proposed changes in the installer to accomplish this can be found int he following PR
https://github.com/openshift/installer/pull/6288/files#diff-26df287c8be6151b827ee3de6fd24366d8736d40089dafdbf7b403d20de22c57